app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,73 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+  def google_oauth2
+
+    auth = request.env["omniauth.auth"]
+
+    if auth.nil? or !auth.has_key?(:provider) or 
+      !auth.has_key?(:uid) or !auth.has_key?(:info) or
+      !auth[:info].has_key?(:email) or !auth[:info].has_key?(:name) or 
+      auth.uid.length == 0 or auth.provider.length == 0 or
+      auth.info.email.length == 0 or auth.info.email.length == 0
```

Comment:
This condition looks really complex. I think it will be better if it is extracted into a separate private function